java/ql/lib/semmle/code/java/security/ImplicitPendingIntents.qll

```diff
@@ -0,0 +1,120 @@
+ *
+ * Extend this class to add additional taint steps that should apply to flows related to the use
+ * of implicit `PendingIntent`s.
+ */
+class ImplicitPendingIntentAdditionalTaintStep extends Unit {
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for flows related to the use of implicit `PendingIntent`s.
+   */
+  predicate step(DataFlow::Node node1, DataFlow::Node node2) { none() }
+
+  /**
+   * Holds if the step from `node1` to `node2` should be considered a taint
+   * step for flows related to the use of implicit `PendingIntent`s. This step is only applicable
+   * in `state1` and updates the flow state to `state2`.
+   */
+  predicate step(
+    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
+    DataFlow::FlowState state2
+  ) {
+    none()
+  }
+}
+
+private class IntentCreationSource extends ImplicitPendingIntentSource {
+  IntentCreationSource() {
+    exists(ClassInstanceExpr cc |
+      cc.getConstructedType() instanceof TypeIntent and this.asExpr() = cc
+    )
+  }
+}
+
+private class SendPendingIntent extends ImplicitPendingIntentSink {
+  SendPendingIntent() {
+    sinkNode(this, "intent-start") and
+    // implicit intents can't be started as services since API 21
+    not exists(MethodAccess ma, Method m |
+      ma.getMethod() = m and
+      m.getDeclaringType().getASupertype*() instanceof TypeContext and
+      m.getName().matches(["start%Service%", "bindService%"]) and
+      this.asExpr() = ma.getArgument(0)
+    )
+    or
+    sinkNode(this, "pending-intent-sent")
+  }
+
+  override predicate hasState(DataFlow::FlowState state) { state = "MutablePendingIntent" }
+}
+
+private class PendingIntentAsFieldAdditionalTaintStep extends ImplicitPendingIntentAdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(Field f |
+      f.getType() instanceof PendingIntent and
+      node1.(DataFlow::PostUpdateNode).getPreUpdateNode() =
+        DataFlow::getFieldQualifier(f.getAnAccess().(FieldWrite)) and
+      node2.asExpr().(FieldRead).getField() = f
+    )
+  }
+}
+
+private class MutablePendingIntentFlowStep extends PendingIntentAsFieldAdditionalTaintStep {
+  override predicate step(
+    DataFlow::Node node1, DataFlow::FlowState state1, DataFlow::Node node2,
+    DataFlow::FlowState state2
+  ) {
+    state1 = "" and
+    state2 = "MutablePendingIntent" and
+    exists(PendingIntentCreation pic, Argument flagArg |
+      node1.asExpr() = pic.getIntentArg() and
+      node2.asExpr() = pic and
+      flagArg = pic.getFlagsArg()
+    |
+      // API < 31, PendingIntents are mutable by default
+      not TaintTracking::localExprTaint(any(ImmutablePendingIntentFlag flag).getAnAccess(), flagArg)
+      or
+      // API >= 31, PendingIntents need to explicitly set mutability
+      TaintTracking::localExprTaint(any(MutablePendingIntentFlag flag).getAnAccess(), flagArg)
```

Comment:
```suggestion
      // We err on the side of false positives here, assuming a PendingIntent may be mutable unless it is at
      // least sometimes explicitly marked immutable and never marked mutable.
      // Note for API level < 31, PendingIntents were mutable by default, whereas since then they are immutable by default.
      not TaintTracking::localExprTaint(any(ImmutablePendingIntentFlag flag).getAnAccess(), flagArg)
      or
      TaintTracking::localExprTaint(any(MutablePendingIntentFlag flag).getAnAccess(), flagArg)
```

To avoid giving the impression we're actually trying to detect the API level in use